x/assets/types/genesis.go

```diff
@@ -207,52 +181,142 @@ func (gs GenesisState) Validate() error {
-			if !info.WaitUnbondingAmount.IsZero() {
+
+			// check for negative values.
+			if info.TotalDepositAmount.IsNegative() || info.WithdrawableAmount.IsNegative() ||
+				info.PendingUndelegationAmount.IsNegative() {
 				return errorsmod.Wrapf(
 					ErrInvalidGenesisData,
-					"non-zero unbonding amount for %s: %s",
-					assetID, info.WaitUnbondingAmount,
+					"negative deposit amount for %s: %+v",
+					assetID, info,
 				)
 			}
-			// check for negative values.
-			if info.TotalDepositAmount.IsNegative() || info.WithdrawableAmount.IsNegative() {
+
+			if info.TotalDepositAmount.GT(tokenTotalStaking) {
 				return errorsmod.Wrapf(
 					ErrInvalidGenesisData,
-					"negative deposit amount for %s: %+v",
+					"invalid deposit amount that is greater than the total staking, assetID: %s: %+v",
 					assetID, info,
 				)
 			}
-			// check that the withdrawable amount and the deposited amount are equal.
-			// this is because this module's genesis only sets up free deposits.
-			// the delegation module bonds them, thereby altering the withdrawable amount.
-			if !info.WithdrawableAmount.Equal(info.TotalDepositAmount) {
+
+			if info.PendingUndelegationAmount.Add(info.WithdrawableAmount).GT(info.TotalDepositAmount) {
 				return errorsmod.Wrapf(
 					ErrInvalidGenesisData,
-					"withdrawable amount is not equal to total deposit amount for %s: %+v",
+					"the sum of PendingUndelegationAmount and WithdrawableAmount is greater than the TotalDepositAmount, assetID: %s: %+v",
 					assetID, info,
 				)
 			}
+			return nil
+		}
+		depositFieldValueF := func(deposit DepositByAsset) (string, struct{}) {
+			return deposit.AssetID, struct{}{}
+		}
+		_, err = utils.CommonValidation(depositByStaker.Deposits, depositFieldValueF, depositValidationF)
+		if err != nil {
+			return errorsmod.Wrap(ErrInvalidGenesisData, err.Error())
+		}
+		return nil
+	}
+	seenFieldValueFunc := func(deposits DepositsByStaker) (string, struct{}) {
+		return deposits.StakerID, struct{}{}
+	}
+	_, err := utils.CommonValidation(gs.Deposits, seenFieldValueFunc, validationFunc)
+	if err != nil {
+		return errorsmod.Wrap(ErrInvalidGenesisData, err.Error())
+	}
+	return nil
+}
+
+// ValidateOperatorAssets performs basic operator assets validation
+func (gs GenesisState) ValidateOperatorAssets(tokensTotalStaking map[string]math.Int) error {
+	validationFunc := func(_ int, assets AssetsByOperator) error {
+		_, err := sdk.AccAddressFromBech32(assets.Operator)
+		if err != nil {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"invalid operator address %s: %s", assets.Operator, err,
+			)
+		}
+
+		// validate the assets list for the specified operator
+		assetValidationFunc := func(_ int, asset AssetByID) error {
+			// check that the asset is registered
+			// no need to check for the validity of the assetID, since
+			// an invalid assetID cannot be in the tokens map.
+			if _, ok := tokensTotalStaking[asset.AssetID]; !ok {
```

Comment:
In assetValidationFunc, tokensTotalStaking[asset.AssetID] can be extracted to avoid redundant map access

> totalStaking, ok := tokensTotalStaking[asset.AssetID]
    if !ok {
...
if asset.Info.TotalAmount.Add(asset.Info.PendingUndelegationAmount).GT(totalStaking) {

Reply:
Thanks, I will fix it in the next commit.